### ISSUE TITLE
Object detection 리서치

### ISSUE THREAD

**Issue by ooakind** [ooakind/apple_grading_system#2]
# Object detection 리서치 내용 정리 이슈
## 목표
평면 사과 이미지에서 결점을 detection 할 수 있는 모델의 설계 방법 및 데이터 형식을 고안하기 위해 다양한 Object detection 알고리즘에 대해 조사한다. 본 이슈는 관련 리서치 내용을 정리한다.

**Comment by ooakind** [comment]
## Face Detection with Deep Learning
[Reference](https://machinelearningmastery.com/how-to-perform-face-detection-with-classical-and-deep-learning-methods-in-python-with-keras/)
### Locating & Localizing

- Localizing: 얼굴의 확장된 영역을 경계표시 하는 것(bounding box) 
- Locating: 얼굴의 좌표를 찾는 것

### Feature-based methods

- Hand-crafted filter를 사용해 face detection
- Classical feature-based techniques(ex.cascade classifier) → OpenCV에서 제공

### Image-based methods

- Learn holistically how to extract faces from the entire image.
- Deep learning을 적용한 결과, Multi-task Cascade Convolutional Neural Network(MTCNN) 개발. → MTCNN 라이브러리에서 제공

## Tutorials

### 1. Cascade Classifier

P. Viola and M. Jones의 논문 Rapid object detection using a boosted cascade of simple features(2001)에서 제안된 방법이다. 

이를 바탕으로 OpenCV 라이브러리를 활용해 face detection 구현 가능하다. 

Pre-trained 모델이 있으며, face가 아닌 다른 데이터셋으로도 학습할 수 있는지 조사가 필요하다.

### 2. MTCNN

Multi-Task Cascaded Convolutional Neural Network, 이하 MTCNN을 활용한 딥러닝 접근 방식으로, Kaipeng Zhang, et al. 의 Joint Face Detection and Alignment Using Multi-task Cascaded Convolutional Network(2016)에서 소개되었다. 

얼굴의 윤곽 및 개수 뿐만 아니라, 눈, 코, 입과 같은 세부적인 요소들도 detection이 가능하며, Tensorflow, OpenCV를 활용해 개발된 MTCNN 라이브러리를 통해 구현할 수 있다. 

사과의 결점을 detection하기 위해 MTCNN을 사용할 필요가 있는지 조사가 필요하다.

**Comment by ooakind** [comment]
## Cascade Classifier 학습(11.8)

### 참고

[OpenCV Haar/cascade training 튜토리얼](https://darkpgmr.tistory.com/70)

[Python Programming Tutorials](https://pythonprogramming.net/haar-cascade-object-detection-python-opencv-tutorial/)

### Training data

positivie, negative 샘플이 모두 필요

- positive: 대상 물체가 포함된 샘플 영상
- negative: 대상 물체가 포함되지 않은 일반 이미지

일단 트레이닝 방법을 정리해보면, 

파라미터를 어떻게 설정하느냐에 따라 데이터 셋을 조정할수도 있고, 필터도 선택할 수 있다. 

설정해야하는 파라미터는 매우 많아보이는데, 사과 결점을 찾아내는 방식으로는 어떤 설정이 가장 적합할 지 찾아내야 한다. 


### 어떻게 cascade classifier를 트레이닝 할 수 있을까?

- OpenCV 사용 방법을 익혀야겠다. cpp을 다뤄야만 사용할 수 있는건지, 파이썬 인터페이스를 제공한다는 게 어디까지 제공한다는건지 등등
- 트레이닝 서버
    - 트레이닝이 얼마나 걸릴지 모르겠다. 
    - positive 이미지의 개수와 크기가 많아질수록 학습 시간이 크게 증가하는 것으로 보인다.
        - Haar Cascade 기준 pos 50000, neg 25000 장 학습 시 일주일
        - 더 빠른 Cascade classifier의 경우 어떨까?
    - 로컬에서 하는 건 한계가 있고, 서버가 필요하긴 할 것 같다. 서버 호스팅을 해야할지, 내 PC에 가상머신 설치해서 사용할지? 윈도우 컴을 서버로 사용한다면?
- pos 이미지의 크기
    - 사과의 결점은 다양한 크기를 가지는데, 사진에서 찾을 수 있는 해당 이미지의 크기 제한이 있는 것으로 보인다. 제한을 늘리려면 학습 시간이 오래 걸리는 것 같은데, 알아볼 필요가 있음.

### 데이터셋을 어떻게 만들까?

- 일단 만들어뒀던 사과 샘플들이 있다. 이걸 사과 배경부분만 꽉 채워서 만들고, 결점을 따로 이미지 화 해야한다. 이미지로 학습하는게 맞나? '영상'이라고 했는데, 정확히 어떤 의미인지 알아봐야 한다.
- Kaggle에서 dataset을 검색해본다.
    - apple로 검색해서 나오는 결과 395가지.
        - 일단 가장 가까운 데이터셋을 구했다. 하지만 cascade classifier를 학습시키려면 일일이 다 가공해야한다.
- positive:
    - (1) 오브젝트만 있는 사진, (2) 오브젝트가 포함된 사진에 그 오브젝트의 위치 (ROI)를 표시) 모두 상관 없다.
    - 그러나 사용하는 positive 이미지가 많아질수록 ROI를 직접 표시하기는 어려워진다. opencv create_samples가 (1)을 사용했을 때 이러한 과정을 대신해주지만, 실제 사진이 아니므로 자연스럽지 않을 수 있다.
- positive vs negative 아이디어
    - positive: 결점 확대 사진, negative: 펼친 사과 배경 사진(결점을 지운 상태)
        - 가장 의도에 가깝다.
        - 펼친 사과를 꼭 사용해야할까? 그냥 둥근 사과 이미지를 neg로 쓰면 어떻게 될까? 일단 다 펼치는건 사실상 불가능하기 때문에 그냥 둥근사과 이미지로 사용해보자.
        - pos는 수작업으로 종류별로 구성해놓자.
    - positive: 결점 확대 사진, negative: 일반 사과 사진
        - negative 이미지가 배경이 있고, 구형이어도 구분이 될까? 한번에 학습이 된다면, 결점과 사과의 꼭지도 구분이 될까?

cascade classifier 사용 방법을 좀 더 구체적으로 계획한 다음 데이터 전처리를 시작하자. 개발 환경은 일단 xml 파일을 만들어야하니까 로컬에서 해야할까? cpp 빌드와 exe 실행을 모두 colab에서 할 수 있나?

- Colab에서 가능하다면 colab에서 데이터 셋 불러오고, exe 실행해서 xml파일 만들고, 그 파일로 cascade classifier 생성해서 결과 출력까지 할 수 있다.
- 불가능하다면 트레이닝 과정(xml 파일 생성)까지는 로컬에 opencv를 설치해서 xml 파일 만들고, 이후부터는 colab 사용

**Comment by ooakind** [comment]
# Object Detection 방법 정리(정리 중)

## Haar Cascade

- 가장 전통적인 Object detection 알고리즘
- 정면으로 보여지는 물체에 대해서는 학습이 잘 되는데, 각도에 따라 모습이 달라지는 경우 성능이 떨어진다.
    - 사과의 결점은 얼마나 각도에 영향을 받을까?

## R-CNN
mask R-cnn, faster r-cnn 등 여러가지 종류가 있다.
- 속도가 느리다. GPU를 반드시 사용해야 한다.

## YOLO
- Real-time R-CNN을 지향하며 제안되었기 때문에 속도가 빠르다.
- 그러나 영상을 7x7의 그리드셀로 분할하고, 각 그리드 셀을 중심으로 하는 오브젝트에 대해서 bounding box 후보를 2개 예측하는 방식이기 때문에 제안하는 후보 수가 훨씬 작아(R-CNN: 1천개 이상, YOLO: 98개) 성능이 떨어진다. 

## MTCNN
- 3개의 네트워크를 합쳐서 학습하는 방식
- Face detection에 특화되어 있는 것인지, 다른 object에 적용한 사례를 찾을 수 없음


## 조사해봐야 할 점
- [ ] 각 방법 별로 Dataset을 어떤 형태로 구축해야하는지 조사
- [ ] 각 방법 별 Real-time 추론이 어느정도로 가능한지 수치적으로 조사
- [ ] PyTorch 기반 faster r-cnn 구현의 경우 속도가 얼마나 향상되었는지 조사
